### ISSUE TITLE
How Can I open a new SSH connection for transferring files using SCP?

### ISSUE THREAD

**Issue by ArcticGene** [ronf/asyncssh#330]
Hi
I wrote the below piece of code, I'm trying to do two tasks on the server, I want to run the "date" command every 0.1 second, and also transfer a file using SCP, but there is a problem, when the two tasks run together, the speed of the date command goes down, instead of every 0.1 second, it runs about every 1 second, but as soon as the file transferred, the speed goes normal.

Does open a new ssh connection solve the problem? I tried to use the create_connection() method, but I didn't manage to.

```
import asyncio, asyncssh, sys


class MySSH:
    def __init__(self):
        pass
        
    async def connect(self):
        host = '127.0.0.1'
        port = 2222
        username = 'user'
        password = 'pass'
        timeout = 5
        try:
            self.ssh_conn = await asyncio.wait_for(asyncssh.connect(
                                                   host, 
                                                   port=port, 
                                                   username=username, 
                                                   password=password, 
                                                   known_hosts=None), 
                                                   timeout=timeout)
            print('SSH connection has been established!')
        except Exception as exc:
            print(exc)

    async def date(self):
        while True:
            res = await self.run_command('date')
            print(res, end='')
            await asyncio.sleep(0.1)
    
    async def transfer(self):
        await asyncssh.scp(dstpath=(self.ssh_conn, '/home/username'), 
                           srcpaths='/home/username/file.iso',
                           progress_handler=self.prgs)

    async def run_command(self, cmd):
        result = await self.ssh_conn.run(cmd)
        return result.stdout
    
    
    def prgs(self, a ,b ,c , d):
        if (c/d) * 100 == 100.0:
            print('File has been transferred')
        
        


async def start():
    ssh = MySSH()
    await ssh.connect()
    task1 =  asyncio.create_task(ssh.date())
    task2 =  asyncio.create_task(ssh.transfer())
    await asyncio.gather(task1, task2)
    
    

asyncio.run(start())
        

```

Here's the output, as you can see, as soon as the file transferred, the speed reaches to 0.1 second.

```
SSH connection has been established!
Sat Nov 14 21:15:05 EST 2020
Sat Nov 14 21:15:05 EST 2020
Sat Nov 14 21:15:06 EST 2020
Sat Nov 14 21:15:07 EST 2020
Sat Nov 14 21:15:08 EST 2020
Sat Nov 14 21:15:09 EST 2020
File has been transferred
Sat Nov 14 21:15:10 EST 2020
Sat Nov 14 21:15:10 EST 2020
Sat Nov 14 21:15:10 EST 2020
Sat Nov 14 21:15:10 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:11 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:12 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:13 EST 2020
Sat Nov 14 21:15:14 EST 2020
Sat Nov 14 21:15:14 EST 2020
Sat Nov 14 21:15:14 EST 2020
Sat Nov 14 21:15:14 EST 2020
Sat Nov 14 21:15:14 EST 2020
Sat Nov 14 21:15:14 EST 2020
```

**Comment by ronf** [comment]
Thanks for the detailed example -- I was able to reproduce what you're seeing here.

I believe what's going on here is that the SCP transfer doesn't yield back to the asyncio event loop until it has completely filled its channel's window. While it calls drain() after every write, there isn't an opportunity to schedule the running of the date command until the writing actually gets paused. On my system here, it looks like it ends up outputting about 2 Mbytes of data before this happens, and that takes some time. Until that completes, the date command doesn't get a chance to run even if the 0.1 second sleep is done.

A potential fix for this would be to give up control after every send. This would probably slow down the transfer a bit due to the extra context switching, but it should provide better interactivity if you are trying to do a transfer like this over the same connection where you are running other commands.

I tried out the following change and it seemed to fix the issue. Does it work for you?

```patch
--- a/asyncssh/scp.py
+++ b/asyncssh/scp.py
@@ -23,6 +23,7 @@
 """SCP handlers"""
 
 import argparse
+import asyncio
 import posixpath
 from pathlib import PurePath
 import shlex
@@ -229,6 +230,7 @@ class _SCPHandler:
 
         self._writer.write(data)
         await self._writer.drain()
+        await asyncio.sleep(0)
 
     def send_ok(self):
         """Send an SCP OK response"""
```

I'm a bit torn about whether to do this unconditionally or not, but I could always add in a parameter to enable/disable this behavior if it causes too much of a performance hit.

**Comment by ArcticGene** [comment]
Thanks Dear Ron,
I tested your solution, it works properly.
I'll be really nice  if you add a parameter for enabling/disabling, this behavior.

What about the other solutions for this?
Can I run the SCP transfer on another thread or possibly create a connection from the main connection and transfer the file from another channel.
I don't have a clear idea, how to do this. I'll be thankful if you help me with this.

**Comment by ronf** [comment]
Thanks for the confirmation. I'm in the middle of some other changes in AsyncSSH at the moment, but I'll look into this in more detail once these other changes are in.

As for other solutions, you're already using multiple channels in the current code. In fact, you're actually opening and closing a new channel each time you run the "date" command, though all the channels share a single SSH connection. Even going with multiple connections wouldn't really help here if you used a single event loop, as you'd still end up with multiple writes being done back to back until the window was full on the SCP, blocking any tasks attempting to use the other connection.

Using threads for this could potentially provide some other benefits around being able to leverage multiple CPU cores on modern processors. One of the downsides of Python asyncio is that you can only effectively use a single core when you have a single event loop, and even going with multiple event loops still limits how much you can take advantage of multiple cores due to the Python global interpreter lock. You end up needing to spawn multiple independent processes on the system each running their own instance of the Python interpreter to really gain the full benefit from that. So, while using multiple threads or processes is an option, you're probably no longer looking at a typical "async" design at that point, and you'd probably need to open multiple independent SSH connections. That might not be so bad, though, if you limited things to two connections, one for the bulk transfer and the other for interactive commands.

Can you say a bit more about what you're really trying to do here? If it is that you want to minimize latency for some interactive SSH requests while having bulk file transfers running in parallel, the option to yield to the event loop during the transfers is probably the best approach.

**Comment by ArcticGene** [comment]
Thank you for your detailed comment. I really appreciate your help.

> If it is that you want to minimize latency for some interactive SSH requests while having bulk file transfers running in parallel, the option to yield to the event loop during the transfers is probably the best approach.

That's it, I exactly want this functionality. What can I do in practice to achieve it?

**Comment by ronf** [comment]
In that case, the change above should provide that, and I'll try to get it into an official release soon.

**Comment by ArcticGene** [comment]
I think, you should take a look at the SFTP client too, I think it needs some changes too, it could be more faster than it is.

**Comment by ronf** [comment]
SFTP shouldn't run into quite the same issue, as it doesn't write large blocks of data from a single coroutine without waiting for other activity. It makes separate requests to the server for each block that it writes, and gets its speed from issuing multiple requests in parallel in different coroutines. However, you can tune the number of such parallel requests with a "max_requests" argument, and reducing that should make it easier for other requests to get through while a transfer is in progress. The downside is that the transfer may take longer when you reduce that, so you'll need to experiment to find a happy medium.

In a quick test here with SCP, adding the sleep(0) added about 5% overhead compared with the current code. That's higher than I'd like, but it did vastly improve the interactivity when doing something in parallel.

I'm torn about adding another config parameter for this, particularly to SCP which is designed to be pretty basic in terms of config settings. Since SFTP can already be made faster than SCP by simply setting a large block size on it, I think letting SCP default to providing the better interactivity and relying on the existing SFTP config settings (block_size and max_requests) is probably the best option here.

**Comment by ArcticGene** [comment]
I tested the SFTP client again, you are right, using the max_requests parameter is a big help.
So you are not going to add a parameter to config the SCP client?

**Comment by ronf** [comment]
Yeah, for now I've made the sleep(0) after every block unconditional for scp. So, without doing anything, you should get the better behavior in the next release when mixing scp transfers with interactive sessions, at the expense of it being slightly slower (about 5% in some testing I did here, though that was hardly exhaustive).

If this turns out to be a problem for anyone, I'll consider making it configurable, but I'd still probably default to this being enabled.

**Comment by ArcticGene** [comment]
Thanks Dear Ron.
I really appreciate your help and effort in developing this useful library.

Best.

**Comment by ronf** [comment]
This change is now released in AsyncSSH 2.5.0. Please let me know if you run into any problems with it.
